core/trino-main/src/main/java/io/trino/operator/window/matcher/ThreadEquivalence.java

```diff
@@ -0,0 +1,248 @@
+        checkArgument(pointer >= 0 && pointer < reachableLabels.size(), "instruction pointer out of program bounds");
+
+        if (firstThread == secondThread || firstLabels.length() == 0) {
+            return true;
+        }
+
+        // compare resulting positions for input navigations
+        Set<LogicalIndexNavigation> distinctPositionsToCompare = new HashSet<>();
+        for (int label : reachableLabels.get(pointer)) {
+            distinctPositionsToCompare.addAll(positionsToCompare.get(label));
+        }
+        for (LogicalIndexNavigation navigation : distinctPositionsToCompare) {
+            if (resolvePosition(navigation, firstLabels) != resolvePosition(navigation, secondLabels)) {
+                return false;
+            }
+        }
+
+        // compare resulting labels for `CLASSIFIER` navigations
+        Set<LogicalIndexNavigation> distinctLabelPositionsToCompare = new HashSet<>();
+        for (int label : reachableLabels.get(pointer)) {
+            distinctLabelPositionsToCompare.addAll(labelsToCompare.get(label));
+        }
+        for (LogicalIndexNavigation navigation : distinctLabelPositionsToCompare) {
+            int firstPosition = resolvePosition(navigation, firstLabels);
+            int secondPosition = resolvePosition(navigation, secondLabels);
+            if ((firstPosition == -1) != (secondPosition == -1)) {
+                return false;
+            }
+            if (firstPosition != -1 && firstLabels.get(firstPosition) != secondLabels.get(secondPosition)) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    private static int resolvePosition(LogicalIndexNavigation navigation, ArrayView labels)
+    {
+        return navigation.resolvePosition(labels.length() - 1, labels, 0, labels.length(), 0);
+    }
+
+    private static List<Set<Integer>> computeReachableLabels(Program program)
+    {
+        List<Set<Integer>> labels = new ArrayList<>(program.size());
+        for (int i = 0; i < program.size(); i++) {
+            labels.add(null);
+        }
+
+        // for every instruction, determine the set of labels reachable from this instruction
+        int computed = 0;
+        while (computed < program.size()) {
+            for (int i = 0; i < program.size(); i++) {
+                if (labels.get(i) != null) {
+                    // already computed
+                    continue;
+                }
+                Instruction instruction = program.at(i);
+                switch (instruction.type()) {
+                    case MATCH_LABEL:
+                        labels.set(i, ImmutableSet.of(((MatchLabel) instruction).getLabel()));
+                        computed++;
+                        break;
+                    case JUMP:
+                        Set<Integer> targetLabels = labels.get(((Jump) instruction).getTarget());
+                        if (targetLabels != null) {
```

Comment:
Wouldn't this require computing the labels for the jump target before we set them here?

Reply:
Yes, the value for `JUMP` is set only if the value for the jump target is already set.

It made me realize that cycles are not handled correctly. I need to fix it.

Reply:
But what if it's not already set? Wouldn't that be an incorrect computation for the reachable labels for the JUMP instruction? In particular, if it's a jump back, it would be set, but if it's a jump forward, it would not.

Reply:
It would work in that case, because it iterates until reachable labels are set for every instruction.
However, it does not account for cycles, so I need to change it.